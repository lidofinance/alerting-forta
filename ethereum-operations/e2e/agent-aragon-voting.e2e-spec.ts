import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-aragon-voting e2e tests", () => {
  let runBlock: (
    blockHashOrNumber: string | number,
    initBlock?: number
  ) => Promise<Finding[]>;
  let runTransaction: (txHash: string) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(
        provideAgentPath("subagents/aragon-voting/agent-aragon-voting")
      ),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should process tx with started vote",
    async () => {
      const findings = await runTransaction(
        "0x69987bf8c4352c40e0429c8492d4842011071524171cd382ea7327d808b37858"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process tx with executed vote",
    async () => {
      const findings = await runTransaction(
        "0x4cc1911b3016ceec169db5b73714b02ee155fb03b6e018fc66a1063a9c1e15fa"
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );

  it(
    "should process block with changed outcome",
    async () => {
      const findings = await runBlock(16691599, 16691598);
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT
  );
});
