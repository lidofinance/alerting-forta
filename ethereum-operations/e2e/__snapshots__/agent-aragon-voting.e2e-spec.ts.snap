// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-aragon-voting e2e tests should process block with changed outcome 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "ARAGON-VOTE-OUTCOME-CHANGED",
  "description": "Expected aragon vote 152 outcome changed from 'Fail' to 'Pass' and there is less than 2 hour(s) left till the end of the first voting phase.",
  "labels": Array [],
  "metadata": Object {},
  "name": "‚ö†Ô∏è Expected vote outcome has changed",
  "protocol": "ethereum",
  "severity": 4,
  "type": 2,
}
`;

exports[`agent-aragon-voting e2e tests should process tx with executed vote 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "ARAGON-VOTE-EXECUTED",
  "description": "Aragon vote 22 was executed",
  "labels": Array [],
  "metadata": Object {
    "args": "22",
  },
  "name": "‚úÖ Aragon: Vote executed",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-aragon-voting e2e tests should process tx with started vote 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "ARAGON-VOTE-STARTED",
  "description": "Aragon vote 30 was started by 0xf73a1260d222f447210581DDf212D915c09a3249
Details:
",
  "labels": Array [],
  "metadata": Object {
    "args": "30,0xf73a1260d222f447210581DDf212D915c09a3249,",
  },
  "name": "üöÄ Aragon: Vote started",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;
